File: experiment/backends/android/inmemory/InMemoryContextManager.kt
// Copyright 2026 Ivan Aliaga <ivan.aliaga@urp.edu.pe>
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package dev.dazzle.experiment

/**
 * In-memory [StorageBackend] using plain Kotlin collections.
 *
 * This is the "no database at all" baseline for the paper. It stores
 * everything in [HashMap]s and [ArrayList]s — zero persistence, zero
 * overhead, zero library dependency. The purpose is to show that raw
 * retrieval latency is dominated by computation (context block string
 * building), not by the storage engine, and that Valkey's value
 * proposition is its capabilities (streams, sorted sets, HFE, etc.)
 * rather than raw speed.
 *
 * Expected result: fastest retrieval of all backends, identical
 * recall / FPR / synthesis because the context block is byte-identical.
 *
 * Limitations vs Valkey:
 *   - No persistence — crash = data loss
 *   - No auto-trim — we manual-trim in code (same as SQLite)
 *   - No atomic operations — single-threaded only
 *   - No TTL / expiration
 *   - No range queries on score — we linear-scan the anomaly list
 *   - No pub/sub, scripting, transactions
 */
class InMemoryContextManager : StorageBackend {

    override val backendName: String = "InMemory"

    // Data structures that mirror Valkey's primitives
    private val readings     = ArrayList<ReadingEntry>(210)
    private val stats        = HashMap<String, Double>()
    private val anomalyMins  = ArrayList<Int>()              // sorted by minute
    private val decisions    = ArrayList<String>()            // ordered by CP index
    private val checkpoints  = HashMap<Int, CheckpointEntry>()

    private data class ReadingEntry(
        val minute: Int,
        val temp: Double,
        val humidity: Double,
        val anomalous: Boolean,
    )

    private data class CheckpointEntry(
        val minute: Int,
        val anomaly: Boolean,
        val severity: String,
        val trend: String,
    )

    // ── Lifecycle ─────────────────────────────────────────────────────────

    override fun flush() {
        readings.clear()
        stats.clear()
        anomalyMins.clear()
        decisions.clear()
        checkpoints.clear()
    }

    // ── Ingest ────────────────────────────────────────────────────────────

    override fun ingest(reading: SensorReading) {
        readings.add(ReadingEntry(
            minute    = reading.minute,
            temp      = reading.tempC,
            humidity  = reading.humidity,
            anomalous = reading.anomalous,
        ))

        // Trim to 200 — equivalent to Valkey MAXLEN ~200
        if (readings.size > 210) {
            readings.subList(0, readings.size - 200).clear()
        }

        // Running stats
        stats["temp_sum"] = (stats["temp_sum"] ?: 0.0) + reading.tempC
        stats["count"]    = (stats["count"] ?: 0.0) + 1.0
        stats["latest_temp"]   = reading.tempC
        stats["latest_minute"] = reading.minute.toDouble()

        val curMin = stats["min_temp"]
        if (curMin == null || reading.tempC < curMin) {
            stats["min_temp"] = reading.tempC
        }
        val curMax = stats["max_temp"]
        if (curMax == null || reading.tempC > curMax) {
            stats["max_temp"] = reading.tempC
        }

        if (reading.anomalous) {
            if (reading.minute !in anomalyMins) {
                // Insert sorted
                val idx = anomalyMins.binarySearch(reading.minute).let {
                    if (it < 0) -(it + 1) else it
                }
                anomalyMins.add(idx, reading.minute)
            }
            stats["anomaly_count"] = (stats["anomaly_count"] ?: 0.0) + 1.0
        }
    }

    // ── Context block (byte-identical to Valkey/SQLite) ───────────────────

    override fun buildContextBlock(currentMinute: Int, windowMinutes: Int): String {
        val lines = mutableListOf<String>()

        // Last 10 readings (oldest → newest)
        val recent = readings.takeLast(10)
        val recentTemps = recent.map { it.temp }

        if (recentTemps.isNotEmpty()) {
            val formatted = recentTemps.joinToString(", ") { String.format("%.1f", it) }
            lines += "Last ${recentTemps.size} temperatures (oldest→newest, °C): $formatted"
            lines += "Recent trend: ${computeTrend(recentTemps)}"
        }

        // Aggregate stats
        val s = readStats()
        if (s != null) {
            lines += "Aggregate over ${s.count} readings: " +
                "avg=${String.format("%.1f", s.avgTemp)}°C, " +
                "min=${String.format("%.1f", s.minTemp)}°C, " +
                "max=${String.format("%.1f", s.maxTemp)}°C"
            lines += "Total anomalies detected so far: ${s.anomalyCount}"
        }

        // Anomalies in window
        val windowStart = maxOf(0, currentMinute - windowMinutes)
        val windowAnomalies = anomalyMins.filter { it in windowStart..currentMinute }

        if (windowAnomalies.isEmpty()) {
            lines += "No anomalies in the last $windowMinutes minutes."
        } else {
            lines += "Anomalous time indices in the last $windowMinutes minutes " +
                "(minute numbers, not temperatures): [${windowAnomalies.joinToString(", ")}]"
        }

        return lines.joinToString("\n")
    }

    override fun buildSynthesisContext(): String {
        val lines = mutableListOf<String>()

        val s = readStats()
        if (s != null) {
            lines += "=== Full Session Stats ==="
            lines += "Total readings: ${s.count}"
            lines += "Temperature range: ${String.format("%.1f", s.minTemp)}°C to " +
                "${String.format("%.1f", s.maxTemp)}°C " +
                "(avg ${String.format("%.1f", s.avgTemp)}°C)"
            lines += "Total anomalies detected: ${s.anomalyCount}"
        }

        if (anomalyMins.isNotEmpty()) {
            lines += "Anomalous time indices (minute numbers, not temperatures): " +
                "[${anomalyMins.joinToString(", ")}]"
        }

        if (decisions.isNotEmpty()) {
            lines += "=== Monitoring Agent Decisions ==="
            for ((idx, decision) in decisions.withIndex()) {
                lines += "  Checkpoint ${idx + 1}: $decision"
            }
        }

        return lines.joinToString("\n")
    }

    // ── Decision storage ──────────────────────────────────────────────────

    override fun storeCheckpointDecision(
        index: Int,
        minute: Int,
        anomalyDetected: Boolean,
        severity: String,
        trend: String,
    ) {
        val decision = "anomaly=${if (anomalyDetected) "yes" else "no"} " +
            "severity=$severity trend=$trend"

        checkpoints[index] = CheckpointEntry(
            minute  = minute,
            anomaly = anomalyDetected,
            severity = severity,
            trend   = trend,
        )

        // Ensure decisions list is large enough
        while (decisions.size <= index) decisions.add("")
        decisions[index] = decision
    }

    // ── Private helpers ───────────────────────────────────────────────────

    private data class RunningStats(
        val count: Int,
        val avgTemp: Double,
        val minTemp: Double,
        val maxTemp: Double,
        val anomalyCount: Int,
    )

    private fun readStats(): RunningStats? {
        val count = stats["count"]?.toInt() ?: return null
        if (count == 0) return null
        val sum      = stats["temp_sum"] ?: 0.0
        val minTemp  = stats["min_temp"] ?: 0.0
        val maxTemp  = stats["max_temp"] ?: 0.0
        val anomCnt  = stats["anomaly_count"]?.toInt() ?: 0
        return RunningStats(
            count        = count,
            avgTemp      = sum / count,
            minTemp      = minTemp,
            maxTemp      = maxTemp,
            anomalyCount = anomCnt,
        )
    }

    private fun computeTrend(temps: List<Double>): String {
        if (temps.size < 2) return "stable"
        val n = temps.size
        val meanX = (n - 1) / 2.0
        val meanY = temps.average()
        val num = (0 until n).sumOf { (it - meanX) * (temps[it] - meanY) }
        val den = (0 until n).sumOf { (it - meanX) * (it - meanX) }
        val slope = if (den != 0.0) num / den else 0.0
        return when {
            slope >  0.15 -> "increasing"
            slope < -0.15 -> "decreasing"
            else          -> "stable"
        }
    }

    // ── Footprint accounting ──────────────────────────────────────────────

    override val backendSizeMethod: String = "inmemory:struct_estimate"

    /**
     * Conservative byte estimate of the live Kotlin objects this backend
     * keeps. The JVM lays each object behind a 16-byte header on 64-bit
     * ART, plus aligned-to-8 payload, so the per-entry numbers below are
     * lower bounds that ignore HashMap / ArrayList growth slack.
     *
     * Components:
     *   - ReadingEntry    : 16 hdr + 4 minute + 8 temp + 8 humidity + 1 anom + pad ≈ 40
     *   - anomalyMins     : 4 bytes per Int (boxed → ~16 each, but ArrayList<Int>
     *                       holds Integer references; conservatively count 16)
     *   - decisions       : variable, sum of String char counts × 2 + 40 hdr each
     *   - checkpoints     : ~80 per CheckpointEntry incl. severity/trend strings
     *   - stats           : HashMap entry node ~48 + key + value ≈ 80 each
     */
    override fun backendSizeBytes(): Long {
        var total = 0L
        total += readings.size.toLong() * 40L
        total += anomalyMins.size.toLong() * 16L
        total += decisions.sumOf { 40L + it.length.toLong() * 2L }
        total += checkpoints.size.toLong() * 80L
        total += stats.entries.sumOf { 80L + it.key.length.toLong() * 2L }
        return total
    }
}

File: experiment/backends/ios/inmemory/InMemoryContextManager.swift
// Copyright 2026 Ivan Aliaga <ivan.aliaga@urp.edu.pe>
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import Foundation

/// In-memory StorageBackend using plain Swift collections.
///
/// This is the "no database at all" baseline. It stores everything in
/// arrays and dictionaries — zero persistence, zero overhead, zero
/// library dependency. The purpose is to show that raw retrieval latency
/// is dominated by computation (context block string building), not by
/// the storage engine.
final class InMemoryContextManager: StorageBackend {

    let backendName: String = "InMemory"

    private struct ReadingEntry {
        let minute: Int
        let temp: Double
        let humidity: Double
        let anomalous: Bool
    }

    private struct CheckpointEntry {
        let minute: Int
        let anomaly: Bool
        let severity: String
        let trend: String
    }

    private var readings:    [ReadingEntry] = []
    private var stats:       [String: Double] = [:]
    private var anomalyMins: [Int] = []             // sorted by minute
    private var decisions:   [String] = []           // ordered by CP index
    private var checkpoints: [Int: CheckpointEntry] = [:]

    // MARK: - Lifecycle

    func flush() {
        readings.removeAll()
        stats.removeAll()
        anomalyMins.removeAll()
        decisions.removeAll()
        checkpoints.removeAll()
    }

    // MARK: - Ingest

    func ingest(_ reading: SensorReading) {
        readings.append(ReadingEntry(
            minute:    reading.minute,
            temp:      reading.tempC,
            humidity:  reading.humidity,
            anomalous: reading.anomalous
        ))

        // Trim to 200 — equivalent to Valkey MAXLEN ~200
        if readings.count > 210 {
            readings.removeFirst(readings.count - 200)
        }

        // Running stats
        stats["temp_sum"]      = (stats["temp_sum"] ?? 0) + reading.tempC
        stats["count"]         = (stats["count"] ?? 0) + 1
        stats["latest_temp"]   = reading.tempC
        stats["latest_minute"] = Double(reading.minute)

        let curMin = stats["min_temp"]
        if curMin == nil || reading.tempC < curMin! {
            stats["min_temp"] = reading.tempC
        }
        let curMax = stats["max_temp"]
        if curMax == nil || reading.tempC > curMax! {
            stats["max_temp"] = reading.tempC
        }

        if reading.anomalous {
            if !anomalyMins.contains(reading.minute) {
                // Insert sorted via binary search
                let idx = anomalyMins.insertionIndex(of: reading.minute)
                anomalyMins.insert(reading.minute, at: idx)
            }
            stats["anomaly_count"] = (stats["anomaly_count"] ?? 0) + 1
        }
    }

    // MARK: - Context block (byte-identical to Valkey/SQLite)

    func buildContextBlock(currentMinute: Int, windowMinutes: Int = 20) -> String {
        var lines: [String] = []

        // Last 10 readings (oldest -> newest)
        let recent = readings.suffix(10)
        let recentTemps = recent.map { $0.temp }

        if !recentTemps.isEmpty {
            let formatted = recentTemps.map { String(format: "%.1f", $0) }.joined(separator: ", ")
            lines.append("Last \(recentTemps.count) temperatures (oldest\u{2192}newest, \u{00B0}C): \(formatted)")
            lines.append("Recent trend: \(computeTrend(Array(recentTemps)))")
        }

        // Aggregate stats
        if let s = readStats() {
            lines.append("Aggregate over \(s.count) readings: " +
                "avg=\(String(format: "%.1f", s.avgTemp))\u{00B0}C, " +
                "min=\(String(format: "%.1f", s.minTemp))\u{00B0}C, " +
                "max=\(String(format: "%.1f", s.maxTemp))\u{00B0}C")
            lines.append("Total anomalies detected so far: \(s.anomalyCount)")
        }

        // Anomalies in window
        let windowStart = max(0, currentMinute - windowMinutes)
        let windowAnomalies = anomalyMins.filter { $0 >= windowStart && $0 <= currentMinute }

        if windowAnomalies.isEmpty {
            lines.append("No anomalies in the last \(windowMinutes) minutes.")
        } else {
            lines.append("Anomalous time indices in the last \(windowMinutes) minutes " +
                "(minute numbers, not temperatures): [\(windowAnomalies.map(String.init).joined(separator: ", "))]")
        }

        return lines.joined(separator: "\n")
    }

    func buildSynthesisContext() -> String {
        var lines: [String] = []

        if let s = readStats() {
            lines.append("=== Full Session Stats ===")
            lines.append("Total readings: \(s.count)")
            lines.append("Temperature range: \(String(format: "%.1f", s.minTemp))\u{00B0}C to " +
                "\(String(format: "%.1f", s.maxTemp))\u{00B0}C " +
                "(avg \(String(format: "%.1f", s.avgTemp))\u{00B0}C)")
            lines.append("Total anomalies detected: \(s.anomalyCount)")
        }

        if !anomalyMins.isEmpty {
            lines.append("Anomalous time indices (minute numbers, not temperatures): " +
                "[\(anomalyMins.map(String.init).joined(separator: ", "))]")
        }

        if !decisions.isEmpty {
            lines.append("=== Monitoring Agent Decisions ===")
            for (idx, decision) in decisions.enumerated() {
                lines.append("  Checkpoint \(idx + 1): \(decision)")
            }
        }

        return lines.joined(separator: "\n")
    }

    // MARK: - Decision storage

    func storeCheckpointDecision(index: Int, minute: Int, anomalyDetected: Bool, severity: String, trend: String) {
        let decision = "anomaly=\(anomalyDetected ? "yes" : "no") severity=\(severity) trend=\(trend)"

        checkpoints[index] = CheckpointEntry(
            minute:   minute,
            anomaly:  anomalyDetected,
            severity: severity,
            trend:    trend
        )

        // Ensure decisions list is large enough
        while decisions.count <= index { decisions.append("") }
        decisions[index] = decision
    }

    // MARK: - Private helpers

    private struct RunningStats {
        let count: Int
        let avgTemp: Double
        let minTemp: Double
        let maxTemp: Double
        let anomalyCount: Int
    }

    private func readStats() -> RunningStats? {
        guard let countD = stats["count"], countD > 0 else { return nil }
        let c       = Int(countD)
        let sum     = stats["temp_sum"] ?? 0
        let minTemp = stats["min_temp"] ?? 0
        let maxTemp = stats["max_temp"] ?? 0
        let anomCnt = Int(stats["anomaly_count"] ?? 0)
        return RunningStats(
            count:        c,
            avgTemp:      sum / Double(c),
            minTemp:      minTemp,
            maxTemp:      maxTemp,
            anomalyCount: anomCnt
        )
    }

    private func computeTrend(_ temps: [Double]) -> String {
        guard temps.count >= 2 else { return "stable" }
        let n = temps.count
        let meanX = Double(n - 1) / 2.0
        let meanY = temps.reduce(0, +) / Double(n)
        let num = (0..<n).reduce(0.0) { $0 + (Double($1) - meanX) * (temps[$1] - meanY) }
        let den = (0..<n).reduce(0.0) { $0 + pow(Double($1) - meanX, 2) }
        guard den != 0 else { return "stable" }
        let slope = num / den
        if slope > 0.15 { return "increasing" }
        if slope < -0.15 { return "decreasing" }
        return "stable"
    }

    // MARK: - Footprint accounting

    var backendSizeMethod: String { "inmemory:struct_estimate" }

    /// Conservative byte estimate of the live Swift objects this backend
    /// keeps. ARM64 lays each class header at 16 bytes, plus aligned
    /// payload — we count payload only and ignore Swift's malloc bucket
    /// rounding, so the result is a lower bound (matches Android's
    /// `inmemory:struct_estimate` accounting).
    ///
    /// Components (per-entry, in bytes):
    ///   - ReadingEntry       : 8 minute + 8 temp + 8 humidity + 1 anom + pad ≈ 32
    ///   - anomalyMins (Int)  : 8 each
    ///   - decisions (String) : 16 hdr + 2 × utf8 length
    ///   - checkpoints        : ~64 per entry incl. severity/trend strings
    ///   - stats              : ~32 per (key, Double) Dictionary entry
    func backendSizeBytes() -> Int64 {
        var total: Int64 = 0
        total += Int64(readings.count) * 32
        total += Int64(anomalyMins.count) * 8
        total += decisions.reduce(0) { $0 + Int64(16 + $1.utf8.count * 2) }
        total += Int64(checkpoints.count) * 64
        total += stats.reduce(0) { $0 + Int64(32 + $1.key.utf8.count * 2) }
        return total
    }
}

// MARK: - Array binary search helper

private extension Array where Element == Int {
    func insertionIndex(of value: Int) -> Int {
        var lo = startIndex, hi = endIndex
        while lo < hi {
            let mid = (lo + hi) / 2
            if self[mid] < value { lo = mid + 1 } else { hi = mid }
        }
        return lo
    }
}
